test/test_ssl.py

```diff
@@ -0,0 +1,73 @@
+import mock
+import pytest
+from six import b
+from urllib3.util import ssl_
+from urllib3.exceptions import SNIMissingWarning
+
+
+@pytest.mark.parametrize('addr', [
+    '::1',
+    '::',
+    '127.0.0.1',
+    '8.8.8.8',
+    b('127.0.0.1')
+])
+def test_is_ipaddress_true(addr):
+    assert ssl_.is_ipaddress(addr)
+
+
+@pytest.mark.parametrize('addr', [
+    'www.python.org',
+    b('www.python.org')
+])
+def test_is_ipaddress_false(addr):
+    assert not ssl_.is_ipaddress(addr)
+
+
+@pytest.mark.parametrize(
+    ['has_sni', 'server_hostname', 'uses_sni'],
+    [(True, '127.0.0.1', False),
+     (False, 'www.python.org', False),
+     (False, '0.0.0.0', False),
+     (True, 'www.google.com', True),
+     (True, None, False),
+     (False, None, False)]
+)
+def test_context_sni_with_ip_address(monkeypatch, has_sni, server_hostname, uses_sni):
+    monkeypatch.setattr(ssl_, 'HAS_SNI', has_sni)
+
+    sock = mock.Mock()
```

Comment:
Can you use `create_autospec` instead of a bare mock throughout?

Reply:
All done!